protographic/src/operations/message-builder.ts

```diff
@@ -0,0 +1,539 @@
+import protobuf from 'protobufjs';
+import {
+  SelectionSetNode,
+  SelectionNode,
+  FieldNode,
+  GraphQLObjectType,
+  GraphQLType,
+  GraphQLSchema,
+  TypeInfo,
+  isObjectType,
+  isEnumType,
+  getNamedType,
+  InlineFragmentNode,
+  FragmentDefinitionNode,
+  GraphQLOutputType,
+  GraphQLEnumType,
+  FragmentSpreadNode,
+  isInterfaceType,
+  isUnionType,
+  GraphQLInterfaceType,
+  GraphQLUnionType,
+} from 'graphql';
+import { mapGraphQLTypeToProto, ProtoTypeInfo } from './type-mapper.js';
+import { FieldNumberManager } from './field-numbering.js';
+import { graphqlFieldToProtoField } from '../naming-conventions.js';
+import { buildEnumType } from './request-builder.js';
+import { upperFirst, camelCase } from 'lodash-es';
+
+/**
+ * Options for building proto messages
+ */
+export interface MessageBuilderOptions {
+  /** Whether to include comments/descriptions */
+  includeComments?: boolean;
+  /** Root object for adding nested types */
+  root?: protobuf.Root;
+  /** Field number manager for consistent numbering */
+  fieldNumberManager?: FieldNumberManager;
+  /** Map of fragment definitions for resolving fragment spreads */
+  fragments?: Map<string, FragmentDefinitionNode>;
+  /** Schema for type lookups */
+  schema?: GraphQLSchema;
+  /** Set to track created enums (to avoid duplicates) */
+  createdEnums?: Set<string>;
+  /** Custom scalar type mappings (scalar name -> proto type) */
+  customScalarMappings?: Record<string, string>;
+  /** Maximum recursion depth to prevent stack overflow (default: 50) */
+  maxDepth?: number;
+  /** Internal: Current recursion depth */
+  _depth?: number;
+  /** Callback to ensure nested list wrapper messages are created */
+  ensureNestedListWrapper?: (graphqlType: GraphQLOutputType) => string;
+}
+
+/**
+ * Builds a Protocol Buffer message type from a GraphQL selection set
+ *
+ * @param messageName - The name for the proto message
+ * @param selectionSet - The GraphQL selection set to convert
+ * @param parentType - The GraphQL type that contains these selections
+ * @param typeInfo - TypeInfo for resolving field types
+ * @param options - Optional configuration
+ * @returns A protobuf Type object
+ */
+export function buildMessageFromSelectionSet(
+  messageName: string,
+  selectionSet: SelectionSetNode,
+  parentType: GraphQLObjectType | GraphQLInterfaceType | GraphQLUnionType,
+  typeInfo: TypeInfo,
+  options?: MessageBuilderOptions,
+): protobuf.Type {
+  const message = new protobuf.Type(messageName);
+  const fieldNumberManager = options?.fieldNumberManager;
+
+  // First pass: collect all field names that will be in this message
+  const fieldNames: string[] = [];
+  const fieldSelections = new Map<string, { selection: FieldNode; type: GraphQLObjectType | GraphQLInterfaceType }>();
+
+  // Maximum recursion depth to prevent stack overflow
+  const maxDepth = options?.maxDepth ?? 50;
+  const currentDepth = options?._depth ?? 0;
```

Comment:
Magic numbers? Add a comment or put them in a const. We should provide some info why we have a maxDepth of 50 by default.